akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/AbstractLeastShardAllocationStrategy.scala

```diff
@@ -0,0 +1,116 @@
+ */
+
+package akka.cluster.sharding.internal
+
+import java.lang.{ Boolean => JBoolean, Integer => JInteger }
+
+import akka.actor.ActorRef
+import akka.actor.ActorSystem
+import akka.actor.Address
+import akka.annotation.InternalApi
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent.CurrentClusterState
+import akka.cluster.Member
+import akka.cluster.MemberStatus
+import akka.cluster.sharding.ShardCoordinator.ActorSystemDependentAllocationStrategy
+import akka.cluster.sharding.ShardRegion.ShardId
+
+import scala.collection.immutable
+
+/**
+ * Common logic for the least shard allocation strategy implementations
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object AbstractLeastShardAllocationStrategy {
+  import MemberStatus._
+  private val JoiningCluster: Set[MemberStatus] = Set(Joining, WeaklyUp)
+  private val LeavingClusterStatuses: Set[MemberStatus] = Set(Leaving, Exiting, Down)
+
+  type AllocationMap = Map[ActorRef, immutable.IndexedSeq[ShardId]]
+
+  final case class RegionEntry(region: ActorRef, member: Member, shardIds: immutable.IndexedSeq[ShardId])
+
+  implicit object ShardSuitabilityOrdering extends Ordering[RegionEntry] {
+    override def compare(x: RegionEntry, y: RegionEntry): Int = {
+      val RegionEntry(_, memberX, allocatedShardsX) = x
+      val RegionEntry(_, memberY, allocatedShardsY) = y
+      if (memberX.status != memberY.status) {
+        // prefer allocating to nodes that are not on their way out of the cluster
+        val xIsLeaving = LeavingClusterStatuses(memberX.status)
+        val yIsLeaving = LeavingClusterStatuses(memberY.status)
+        JBoolean.compare(xIsLeaving, yIsLeaving)
+      } else if (memberX.appVersion != memberY.appVersion) {
+        // prefer nodes with the highest rolling update app version
+        memberY.appVersion.compareTo(memberX.appVersion)
+      } else {
+        // prefer the node with the least allocated shards
+        JInteger.compare(allocatedShardsX.size, allocatedShardsY.size)
+      }
+    }
+  }
+}
+
+/**
+
+ *
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] abstract class AbstractLeastShardAllocationStrategy extends ActorSystemDependentAllocationStrategy {
+  import AbstractLeastShardAllocationStrategy._
+
+  @volatile private var cluster: Cluster = _
+
+  override def start(system: ActorSystem): Unit = {
+    cluster = Cluster(system)
+  }
+
+  // protected for testability
+  protected def clusterState: CurrentClusterState = cluster.state
+  protected def selfMember: Member = cluster.selfMember
+
+  final protected def isAGoodTimeToRebalance(regionEntries: Iterable[RegionEntry]): Boolean = {
+    // Avoid rebalance when rolling update is in progress
+    // (This will ignore versions on members with no shard regions, because of role or not yet completed joining)
+    val version = regionEntries.head.member.appVersion
+    def allNodesSameVersion = {
+      regionEntries.forall(_.member.appVersion == version)
```

Comment:
that would be more efficient as
```
regionEntries.exists(_.member.appVersion != version)
```
because then it doesn't have to iterate over all entries if it finds a different version

Reply:
Happy case needs to examine all and`.forall` (in `iterator.forall`) bails early as soon as the the predicate is false so I think this is just as efficient.

Reply:
Doh, I don't know what I was thinking